lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -0,0 +1,148 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+-- |
+-- Copyright: © 2020 IOHK
+-- License: Apache-2.0
+--
+-- Haskell-node "shelley" implementation of the @StakePoolLayer@ abstraction,
+-- i.e. some boring glue.
+module Cardano.Wallet.Shelley.Pools where
+
+import Prelude
+
+import Cardano.Wallet.Api.Types
+    ( ApiT (..) )
+import Cardano.Wallet.Network
+    ( ErrNetworkUnavailable, NetworkLayer (..) )
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..), GenesisParameters (..), PoolId )
+import Cardano.Wallet.Shelley.Compatibility
+    ( Shelley, toPoint )
+import Cardano.Wallet.Shelley.Network
+    ( NodePoolLsqData (..) )
+import Cardano.Wallet.Unsafe
+    ( unsafeMkPercentage, unsafeRunExceptT )
+import Control.Monad.IO.Class
+    ( liftIO )
+import Control.Monad.Trans.Except
+    ( ExceptT (..), runExceptT )
+import Data.Map
+    ( Map )
+import Data.Map.Merge.Strict
+    ( dropMissing, traverseMissing, zipWithMatched )
+import Data.Ord
+    ( Down (..) )
+import Data.Quantity
+    ( Percentage (..), Quantity (..) )
+import Data.Sort
+    ( sortOn )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Api.Types as Api
+import qualified Data.Map.Merge.Strict as Map
+import qualified Data.Map.Strict as Map
+
+-- | Stake Pool Data fields fetched from the node via LSQ
+data PoolLsqMetrics = PoolLsqMetrics
+    { nonMyopicMemberRewards :: Quantity "lovelace" Word64
+    , relativeStake :: Percentage
+    , saturation :: Double
+    } deriving (Eq, Show, Generic)
+
+combineLsqData
+    :: NodePoolLsqData
+    -> Map PoolId PoolLsqMetrics
+combineLsqData NodePoolLsqData{nOpt, rewards, stake} =
+    Map.merge stakeButNoRewards rewardsButNoStake bothPresent stake rewards
+  where
+    -- calculate the saturation from the relative stake
+    sat s = fromRational $ (getPercentage s) / (1 / fromIntegral nOpt)
+
+    -- If we fetch non-myopic member rewards of pools using the wallet
+    -- balance of 0, the resulting map will be empty. So we set the rewards
+    -- to 0 here:
+    stakeButNoRewards = traverseMissing $ \_k s -> pure $ PoolLsqMetrics
+        { nonMyopicMemberRewards = Quantity 0
+        , relativeStake = s
+        , saturation = (sat s)
+        }
```

Comment:
Doesn't the spec suggests something about this? Should we order pool randomly in that case?

Reply:
I thought that was when we don't have enough data.

I would think that ranking according to the rewards _if_ the user had money, would be a better thing to do.

A user can't delegate without any money anyway.

We could include ROI in % and return `null` for the non-myopic member rewards.